### PR TITLE
Add timeline query feature for memory retrieval by date

### PR DESCRIPTION
**Pull request by marcelsamyn**
## Summary
This PR adds a new timeline query feature that allows users to retrieve memories grouped by date. The implementation includes a database query function, request/response schemas with validation, comprehensive tests, and SDK integration.

## Key Changes
- **Timeline Query Function** (`src/lib/query/timeline.ts`): Implements efficient batch querying of temporal (day) nodes within a date range, with support for:
  - Date range filtering (defaults to last 90 days)
  - Pagination via limit/offset on days
  - Optional node type filtering on connected nodes
  - Deduplication of nodes connected via multiple edges
  - Single batch query to avoid N+1 query problems

- **Schema Definitions** (`src/lib/schemas/query-timeline.ts`): Zod schemas for request/response validation including:
  - `QueryTimelineRequest`: userId (required), startDate/endDate (YYYY-MM-DD format), limit (1-100, default 30), offset (default 0), nodeTypes (optional array)
  - `QueryTimelineResponse`: days array with date, temporalNodeId, nodeCount, and connected nodes; totalDays and hasMore for pagination

- **Comprehensive Tests** (`src/lib/schemas/query-timeline.test.ts`): Full test coverage for schema validation including edge cases (invalid dates, invalid node types, limit bounds, negative offsets)

- **API Route** (`src/routes/query/timeline.ts`): H3 event handler that validates input and returns validated response

- **SDK Integration** (`src/sdk/memory-client.ts`): Added `queryTimeline()` method to MemoryClient for client-side usage

## Notable Implementation Details
- Uses lexicographic comparison for date filtering since temporal node labels are stored as "YYYY-MM-DD" strings
- Employs SQL CASE statements to handle bidirectional edge relationships (day node can be source or target)
- Counts total unique nodes before applying nodeType filter to provide accurate nodeCount
- Deduplicates nodes at two stages: total count and filtered results

https://claude.ai/code/session_01QnhM7Q1mZf97d5Uj28kcwy

### CODE REVIEW

File: src/lib/query/timeline.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    userId,
+    limit = 30,
+    offset = 0,
+    nodeTypes,
+  } = params;
+
+  const today = format(new Date(), "yyyy-MM-dd");
+  const ninetyDaysAgo = format(subDays(new Date(), 90), "yyyy-MM-dd");
+  const startDate = params.startDate ?? today;
+  const endDate = params.endDate ?? ninetyDaysAgo;
```
[the remark this answers is not in the document]
Comment (marcelsamyn) [reply src/lib/query/timeline.ts:28]: The `startDate` = today / `endDate` = 90 days ago convention is intentional — it was designed by the consuming app (Petals) where "start" means "where the timeline view starts" (today) and "end" means "how far back to look". That said, the naming can be confusing, so I've added date normalization in 06d4193 that sorts the two dates internally. The query now works correctly regardless of which param is chronologically larger, so callers can pass them in either order.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/query/timeline.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    })
+    .from(nodes)
+    .innerJoin(nodeMetadata, eq(nodeMetadata.nodeId, nodes.id))
+    .where(
+      and(
+        eq(nodes.userId, userId),
+        eq(nodes.nodeType, NodeTypeEnum.enum.Temporal),
+        // label is the date string: compare lexicographically
+        lte(nodeMetadata.label, startDate),
+        gte(nodeMetadata.label, endDate),
```
[the remark this answers is not in the document]
Comment (marcelsamyn) [reply src/lib/query/timeline.ts]: The operators were already correct for the original defaults (`lte(label, today)` + `gte(label, 90daysAgo)` = dates between 90 days ago and today). With the normalization added in 06d4193, the query now uses `gte(label, rangeMin)` + `lte(label, rangeMax)` which reads more naturally and works regardless of parameter ordering.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/query/timeline.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      ),
+    )
+    .orderBy(desc(nodeMetadata.label));
+
+  const totalDays = allDayNodes.length;
+
+  // Step 2: Apply pagination to the day nodes.
+  const paginatedDayNodes = allDayNodes.slice(offset, offset + limit);
```
[the remark this answers is not in the document]
Comment (marcelsamyn) [reply src/lib/query/timeline.ts]: Good catch — fixed in 06d4193. Pagination now uses DB-level `LIMIT`/`OFFSET` with a separate `COUNT(*)` query for `totalDays`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/lib/query/timeline.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    }
+    const totalNodeCount = uniqueMap.size;
+
+    // Apply nodeTypes filter if requested (after counting total)
+    let filteredNodes = allConnected;
+    if (nodeTypes && nodeTypes.length > 0) {
+      filteredNodes = allConnected.filter((r) =>
+        nodeTypes.includes(r.nodeType as typeof nodeTypes[number]),
+      );
+    }
```
[the remark this answers is not in the document]
Comment (marcelsamyn) [reply src/lib/query/timeline.ts:167]: The `nodeCount` being pre-filter is intentional per the spec — it lets the Petals UI show context like "5 memories (showing 2 Events)" when filtering. The in-memory filter is acceptable here since per-day connected node counts are small (typically tens, not thousands). Pushing the filter to SQL would require either two separate batch queries or a conditional count expression, adding complexity without meaningful perf gain at this scale.

---
_Generated by [Claude Code](https://claude.ai/code)_
